### ISSUE TITLE
How to impute an ordinal variable with mice but prevent it from taking one value?

### ISSUE THREAD

**Issue by stefvanbuuren** [amices/mice#224]
Laurence Geebelen alerted me that the script on <https://stats.stackexchange.com/questions/32179/how-to-impute-an-ordinal-variable-with-mice-but-prevent-it-from-taking-one-value?newreg=9d716504b05846469c7e42bc0bc18614> does not work in `mice 3.8.0`.

```r
library(mice)
df <- data.frame(age = c(24,37,58,65,70,84),
                 overall_tumor_grade = c(1,1,2,3,'X',NA),
                 tumor_size = c(1.5,2.0,4.2,5.6,0,0.1))

mice.impute.tumor <- function(y, ry, x, ...){
  ymis <- y[!ry]
  tmis <- x$tumor_size[!ry] > 0
  t  <- x$tumor_size > 0
  y[!ry] <- NA
  ymis[!tmis] <- "X"
  ymis[tmis] <- mice.impute.polyreg(y[t, drop = TRUE], ry[t], x[t,], ...)
  ymis
}

ini <- mice(df, maxit = 0)
meth <- ini$meth
meth["overall_tumor_grade"] <- "tumor"
imp <- mice(df, meth = meth, maxit = 1, m = 2)
```

```
 iter imp variable
  1   1  overall_tumor_gradeError: $ operator is invalid for atomic vectors
> 
```

**Comment by prockenschaub** [comment]
Could it be that at some point during the migration to `mice 3.8.0` `x` changed from being passed on as a `data.frame` to being passed on as a `matrix`? The following changes seem to fix the code by using `[` instead of `$`:

``` r
library(mice)
df <- data.frame(age = c(24,37,58,65,70,84),
                 overall_tumor_grade = c(1,1,2,3,'X',NA),
                 tumor_size = c(1.5,2.0,4.2,5.6,0,0.1), 
                 stringsAsFactors = TRUE) # Explicitly define this, since my default.stringsAsFactors = FALSE

mice.impute.tumor <- function(y, ry, x, wy, ...){
  ymis <- y[!ry]
  tmis <- x[, 'tumor_size'][!ry] > 0     # Change `$` to `[`
  t  <- x[, 'tumor_size'] > 0            # Change `$` to `[`
  y[!ry] <- NA
  ymis[!tmis] <- "X"
  ymis[tmis] <- mice.impute.polyreg(y[t, drop = TRUE], ry[t], x[t,], wy[t], ...)
  ymis
}

ini <- mice(df, maxit = 0)
meth <- ini$meth
meth["overall_tumor_grade"] <- "tumor"
imp <- mice(df, meth = meth, maxit = 1, m = 2)
#> 
#>  iter imp variable
#>   1   1  overall_tumor_grade
#>   1   2  overall_tumor_grade
```

In addition, I also added `wy` in the function header because otherwise it would be passed on with 6 element, whereas `y`, `ry` and `x` where passed on with 5 elements/rows.

**Comment by stefvanbuuren** [comment]
Ahh, yes. Great.

I wrote the initial script in the year 2012. Versions prior to `mice 3.0.0` used the now retired `padModel()` function to create the design matrix. That function returned a data frame. Since `mice 3.0.0` we use `base::model.matrix()`, which returns a matrix, so the `$` operator on `x` does not work anymore.

Also, the `wy` argument is new in `mice 3.0.0`, so it is good to add it. 

I added a note to my Stack Exchange response that points to your solution.
